### PR TITLE
Partial CLI-57: EventIndexCache

### PR DESCRIPTION
**Pull request by sygao**
- Add `ref.event_index_cache.EventIndexCache`
- Add `EventSourcedRepo._event_index_cache` and integrate cache and load into default `.get_otu()` operation
- Add `EventSourcedRepo._get_event_index()` and `._get_event_index_after_start()` - - Add `EventSourcedRepo._load_otu_event_index()`

### CODE REVIEW

**Review by igboyes — CHANGES_REQUESTED** [review]

File: virtool_cli/ref/event_index_cache.py [hunk 1]
```diff
@@ -0,0 +1,140 @@
...
+        otu_cache_logger = logger.bind(
+            cached_events=otu_index_cache.events, otu_id=str(otu_id)
+        )
+
+        if otu_index_cache.at_event > last_id:
+            logger.warning(
+                "Index cache does not belong to current repo. Deleting cached data...",
+                cached_at_event=otu_index_cache.at_event,
+                latest_event=last_id,
+            )
```
[the remark this answers is not in the document]
Comment (sygao) [reply virtool_cli/ref/event_index_cache.py:112]: True, this event mostly happens when debugging... 

I do want to handle this problem somehow, but an exception-based solution would be better. `EventCacheException`?

File: virtool_cli/ref/event_index_cache.py [hunk 2]
```diff
@@ -0,0 +1,140 @@
...
+from pydantic.dataclasses import dataclass
+
+from structlog import get_logger
+
+logger = get_logger("repo.cache.event_index")
+
+
+@dataclass
+class OTUEventIndex:
+    at_event: Annotated[int, Field(ge=1)]
```
[the remark this answers is not in the document]
Comment (sygao) [reply virtool_cli/ref/event_index_cache.py:17]: `pydantic.dataclasses` seems to work. Could switch to BaseModel easily if you think it's okay.

[the remark this answers is not in the document]
Comment (igboyes) [reply virtool_cli/ref/event_index_cache.py:18]: Ah. I didn't see the import and have not seen `pydantic.dataclasses` used before. Is there a benefit over `BaseModel`?

File: virtool_cli/ref/event_index_cache.py (hunk 1 above)
[the remark this answers is not in the document]
Comment (igboyes) [reply virtool_cli/ref/event_index_cache.py:109]: Sounds good!

**Comment by sygao** [comment]
My changes were too aggressive to auto-merge, so I had to merge by hand.

I've uploaded the old branch to my [own fork](https://github.com/sygao/virtool-cli/tree/id-event-cache-old) for now.

File: virtool_cli/ref/event_index_cache.py [hunk 3]
```diff
@@ -0,0 +1,140 @@
...
+            UUID(subpath.stem)
+            for subpath in self.path.iterdir()
+            if subpath.suffix == ".json"
+        ]
+
+    def populate(self, event_index: dict[UUID, list[int]], last_id: int):
+        """Takes an event index and caches each individual OTU's events
+        under a separate file."""
+        for otu_id in event_index:
+            self.cache_otu_events(otu_id, event_index[otu_id], last_id=last_id)
```
[the remark this answers is not in the document]
Comment (sygao) [reply virtool_cli/ref/event_index_cache.py:49]: I initially thought that the event index was more like a TOC, and the terminology seems to have gotten muddled. I may do some more general renaming to make this all work.

**Pull request merged**